src/Nethermind/Nethermind.Evm/EOF/EvmObjectFormat.cs

```diff
@@ -284,32 +288,29 @@ public bool ValidateBody(ReadOnlySpan<byte> container, in EofHeader header)
                 return false;
             }
 
-            ReadOnlySpan<byte> typesection = container.Slice(typeSectionStart, typeSectionSize);
-            if (!ValidateTypeSection(typesection))
+            ReadOnlyMemory<byte> typesection = container.Slice(typeSectionStart, typeSectionSize);
+            if (!ValidateTypeSection(typesection.Span))
             {
                 if (Logger.IsTrace) Logger.Trace($"EIP-4750: invalid typesection found");
                 return false;
             }
 
-            for (int sectionIdx = 0; sectionIdx < header.CodeSections.Length; sectionIdx++)
+            int validSection = 0;
+            Parallel.For(0, header.CodeSections.Length, (sectionIdx, state) =>
```

Comment:
So does it provide speed-up in the end? I think the previous summary was that it doesn't?

Reply:
yeah, re-attempting it because with EIP5450 the code section checks became a bit more complex

Reply:
Can you post benchmarks?

Reply:
sure I am working on them (I'll post them as soon as I am done with them )

Reply:
@LukaszRozmej well this is underwhelming, and unexpected
  
source : https://github.com/DemuirGos/Benchmarks/tree/PR-5149-Vs-PR-4950
```
BenchmarkDotNet=v0.13.3, OS=Windows 11 (10.0.22621.1105)
Intel Core i5-8265U CPU 1.60GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
```
|                  Method |      Mean |    Error |    StdDev |    Gen0 | Allocated |
|------------------------ |----------:|---------:|----------:|--------:|----------:|
|   NewWay_Valid_Bytecode | 179.22 us | 3.580 us |  6.271 us | 32.9590 |  79.77 KB |
| NewWay_Invalid_Bytecode | 172.61 us | 3.377 us |  5.453 us | 33.2031 |  79.68 KB |
|   OldWay_Valid_Bytecode |  92.52 us | 2.871 us |  8.419 us |  7.8125 |  24.12 KB |
| OldWay_Invalid_Bytecode |  91.25 us | 3.403 us | 10.034 us |  7.8125 |  24.12 KB |

Reply:
@LukaszRozmej well this is underwhelming, and unexpected
  
source : https://github.com/DemuirGos/Benchmarks/tree/PR-5149-Vs-PR-4950
```
BenchmarkDotNet=v0.13.3, OS=Windows 11 (10.0.22621.1105)
Intel Core i5-8265U CPU 1.60GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
```
|                  Method |      Mean |    Error |    StdDev |    Gen0 | Allocated |
|------------------------ |----------:|---------:|----------:|--------:|----------:|
|   NewWay_Valid_Bytecode | 179.22 us | 3.580 us |  6.271 us | 32.9590 |  79.77 KB |
| NewWay_Invalid_Bytecode | 172.61 us | 3.377 us |  5.453 us | 33.2031 |  79.68 KB |
|   OldWay_Valid_Bytecode |  92.52 us | 2.871 us |  8.419 us |  7.8125 |  24.12 KB |
| OldWay_Invalid_Bytecode |  91.25 us | 3.403 us | 10.034 us |  7.8125 |  24.12 KB |

Reply:
And without parallelization?

Reply:
@LukaszRozmej here are the results, is there any reason why it gets worse : 
```
BenchmarkDotNet=v0.13.3, OS=Windows 11 (10.0.22621.1105)
Intel Core i5-8265U CPU 1.60GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
```
|                             Method |      Mean |    Error |   StdDev |    Gen0 | Allocated |
|----------------------------------- |----------:|---------:|---------:|--------:|----------:|
|      Parallel_Stack_Valid_Bytecode | 231.06 us | 3.473 us | 3.249 us | 66.1621 | 172.84 KB |
|    Parallel_Stack_Invalid_Bytecode | 226.25 us | 3.461 us | 3.237 us | 57.1289 | 153.74 KB |
|     Parallel_Pooled_Valid_Bytecode | 155.19 us | 1.593 us | 1.412 us | 37.8418 |  94.04 KB |
|   Parallel_Pooled_Invalid_Bytecode | 150.54 us | 1.417 us | 1.256 us | 37.3535 |  91.59 KB |
|    Sequential_Stack_Valid_Bytecode |  55.81 us | 0.768 us | 0.718 us |  7.8735 |  24.12 KB |
|  Sequential_Stack_Invalid_Bytecode |  55.79 us | 0.682 us | 0.533 us |  7.8735 |  24.12 KB |
|   Sequential_Pooled_Valid_Bytecode |  38.55 us | 0.747 us | 0.734 us |  5.2490 |  16.09 KB |
| Sequential_Pooled_Invalid_Bytecode |  37.99 us | 0.489 us | 0.408 us |  5.2490 |  16.09 KB |

Reply:
Sequential > Parallel because the work is too small, so creating task and scheduling it to thread has higher overhead than gain. Not sure how many sections your data set has.

Pooled > Stack, as Pool doesn't allocate any of two bitmaps, but stack allocates `SortedSet<int> jumpdests = new();`

Go with Sequential Pool

Reply:
Got it 👍🏼

Reply:
Can you like profile why we are still getting 16KB of memory allocated in `Sequential_Pooled` ?

Reply:
Sure

Reply:
https://www.jetbrains.com/help/dotmemory/Analyze_Memory_Allocation.html

Reply:
from what I found the ValidateStack uses a lot of memory for the Dictionary and other stuff, I am trying to see if I can use a better DS for it

Reply:
@LukaszRozmej I was trying something out, I attempted to remove the Dictionary and Stack instances used in the StackValidation and replace them with Arrays like the following (still a draft): 
* new code : [Code not using Dictionary/Stack](https://github.com/DemuirGos/Benchmarks/blob/11353439b0a16e0e507053dec42032ee812beefc/CodeValidator/BytecodeValidatorPoolSeqOpt.cs#L473-L584)
* old code : [Code using Dictionary/Stack](https://github.com/DemuirGos/Benchmarks/blob/11353439b0a16e0e507053dec42032ee812beefc/CodeValidator/BytecodeValidatorPoolSeqOld.cs#L470-L572)

Benchmark results : 
* benchmark code : https://github.com/DemuirGos/Benchmarks/tree/ParralelForVsFor
```
BenchmarkDotNet=v0.13.3, OS=Windows 11 (10.0.22621.1105)
Intel Core i5-8265U CPU 1.60GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
```

|                                 Method |     Mean |     Error |    StdDev |     Gen0 | Allocated |
|--------------------------------------- |---------:|----------:|----------:|---------:|----------:|
|   Sequential_Pooled_Valid_Bytecode_Old | 7.260 ms | 0.1407 ms | 0.1925 ms | 765.6250 | 2448855 B |
| Sequential_Pooled_Invalid_Bytecode_Old | 3.141 ms | 0.2202 ms | 0.6492 ms | 386.7188 | 1224568 B |
|   Sequential_Pooled_Valid_Bytecode_New | 3.274 ms | 0.1505 ms | 0.4439 ms |        - |     309 B |
| Sequential_Pooled_Invalid_Bytecode_New | 1.713 ms | 0.0907 ms | 0.2659 ms |        - |     302 B |

Reply:
Ah this is the part I haven't reviewed yet :)
Make PushWorklet/PopWorklet static and pass stuff by reference if needed.
Shouldn't `stackHeight`, `jumpDestination` just be `ushort`?

Reply:
Also does the array need to be actually cleared on return?

Reply:
> Shouldn't `stackHeight`, `jumpDestination` just be `ushort`?
I believe they can

> Also does the array need to be actually cleared on return?
well I think I left it there because apparently when I rent a ``ushort?[]`` it gives an array full of garbage (like C heap arrays) but I have to try since now I am using ``ushort[]`` (I think now it give an array full of 0s)
I used ushort?[] first because I had to find a way to encode the three possible state of a cell : visited with stack-height 0, visited with a non-zero stackheight, and not-visited, but now I am using a work around until I find a better alternative, since the stack height can never be negative I consider `0` as ``not visited`` and if it is I store the ``stackHeight+1``  (Ugly I know)

Reply:
It is fine for it to be garbage, as you track your head-index and put non-garbage?